flang/lib/Lower/Bridge.cpp

```diff
@@ -869,226 +881,223 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+      builder->create<fir::StoreOp>(loc, tripCount, info.tripVariable);
+      builder->create<fir::StoreOp>(loc, lowerValue, info.loopVariable);
+
+      // Unstructured loop header - generate loop condition and mask.
+      // Note - Currently there is no way to tag a loop as a concurrent loop.
+      startBlock(info.headerBlock);
+      tripCount = builder->create<fir::LoadOp>(loc, info.tripVariable);
+      auto zero = builder->createIntegerConstant(loc, controlType, 0);
+      auto cond = builder->create<mlir::CmpIOp>(loc, mlir::CmpIPredicate::sgt,
+                                                tripCount, zero);
       if (info.maskExpr) {
-        auto ifOp = builder->create<fir::IfOp>(
-            loc, createFIRExpr(loc, info.maskExpr), /*withElseRegion=*/false);
-        builder->setInsertionPointToStart(&ifOp.thenRegion().front());
+        genFIRConditionalBranch(cond, info.maskBlock, info.exitBlock);
+        startBlock(info.maskBlock);
+        auto latchBlock = getEval().getLastNestedEvaluation().block;
+        assert(latchBlock && "missing masked concurrent loop latch block");
+        genFIRConditionalBranch(createFIRExpr(loc, info.maskExpr),
+                                info.bodyBlock, latchBlock);
+      } else {
+        genFIRConditionalBranch(cond, info.bodyBlock, info.exitBlock);
+        if (&info != &incrementLoopNestInfo.back()) // not innermost
+          startBlock(info.bodyBlock); // preheader block of enclosed dimension
+      }
+      if (!info.localInitSymList.empty()) {
+        auto insertPt = builder->saveInsertionPoint();
+        builder->setInsertionPointToStart(info.bodyBlock);
+        genLocalInitAssignments(info);
+        builder->restoreInsertionPoint(insertPt);
       }
-      genLocalInitAssignments();
-      return;
-    }
-
-    // Unstructured loop preheader - initialize tripVariable and loopVariable.
-    mlir::Value tripCount;
-    if (info.hasRealControl) {
-      auto delta1 = builder->create<mlir::SubFOp>(loc, upperValue, lowerValue);
-      auto delta2 = builder->create<mlir::AddFOp>(loc, delta1, info.stepValue);
-      tripCount = builder->create<mlir::DivFOp>(loc, delta2, info.stepValue);
-      controlType = builder->getIndexType();
-      tripCount = builder->createConvert(loc, controlType, tripCount);
-    } else {
-      auto delta1 = builder->create<mlir::SubIOp>(loc, upperValue, lowerValue);
-      auto delta2 = builder->create<mlir::AddIOp>(loc, delta1, info.stepValue);
-      tripCount =
-          builder->create<mlir::SignedDivIOp>(loc, delta2, info.stepValue);
-    }
-    if (fir::isAlwaysExecuteLoopBody()) { // minimum tripCount is 1
-      auto one = builder->createIntegerConstant(loc, controlType, 1);
-      auto cond = builder->create<mlir::CmpIOp>(loc, CmpIPredicate::slt,
-                                                tripCount, one);
-      tripCount = builder->create<mlir::SelectOp>(loc, cond, one, tripCount);
```

Comment:
These diffs are really difficult to read. 50 lines "erased" followed by 50 lines that look the same ...